File: adapter/src/convert.js
/**
 * KONVERTERING — det ENE stedet enheter byttes (D57, D58, spec §11 krav 5).
 *
 * Motoren regner i yards og mph og skal aldri røres av enhetsvalget (D27).
 * Denne modulen er hele konverteringslaget: finnes det en yard→meter-
 * multiplikasjon utenfor denne fila, er det en feil — og `tools/
 * lint-physics.mjs` feiler på den.
 *
 * D57 (2026-08-25): to enhetspakker.
 *   'meters'  → avstander i meter, fart i mph
 *   'yards'   → avstander i yards, fart i mph
 *
 * Fart er ALLTID mph i begge — golfens universelle enhet for kølle- og
 * ballfart. Enhetsuavhengig og upåvirket av valget: spinn (rpm), smash
 * (forholdstall), vinkler (grader) og Studios mm-verdier.
 *
 * Konverteringsfaktoren importeres fra motorens konstantregister framfor å
 * gjentas som literal her. Det er ikke pynt: linten forbyr yard→meter-
 * literalen utenfor engine/, så faktoren KAN ikke dupliseres uten at `npm test` feiler.
 *
 * Kontrakt som i motoren (spec §3): endelige tall inn, ellers kast.
 * Ingen parsing, ingen koersjon.
 */

import { yardToMetre } from '../../engine/src/constants.js';

/** De to enhetspakkene fra D57. Interne id-er, ikke brukervendt kopi. */
export const UNIT_SYSTEMS = Object.freeze(['meters', 'yards']);

/** @throws {TypeError} når `system` ikke er en av de to pakkene */
export function assertUnitSystem(system) {
  if (!UNIT_SYSTEMS.includes(system)) {
    throw new TypeError(
      `ukjent enhetssystem: ${String(system)}. Gyldige: ${UNIT_SYSTEMS.join(', ')} (D57).`,
    );
  }
}

function assertFinite(value, name) {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new TypeError(
      `${name} må være et endelig tall, fikk ${String(value)}. ` +
        'Parsing og koersjon hører ikke hjemme i adapteren heller.',
    );
  }
}

/**
 * Avstand fra motor (yards) til visningsenhet. Verdien returneres URUNDET —
 * avrunding (D28) skjer i formatlaget, ETTER konvertering, aldri før.
 *
 * @param {number} yards motorens tall, urørt
 * @param {'meters'|'yards'} unitSystem
 * @returns {{value: number, unit: 'm'|'yd'}}
 */
export function distanceForDisplay(yards, unitSystem) {
  assertFinite(yards, 'yards');
  assertUnitSystem(unitSystem);
  if (unitSystem === 'yards') return { value: yards, unit: 'yd' };
  return { value: yards * yardToMetre, unit: 'm' };
}

/**
 * Fart. D57: alltid mph, uansett enhetspakke. Funksjonen finnes for at
 * kallstedene skal gå gjennom konverteringslaget også der svaret er
 * identitet — så regelen står ett sted, ikke implisitt i hvert kallsted.
 *
 * @param {number} mph motorens tall, urørt
 * @returns {{value: number, unit: 'mph'}}
 */
export function speedForDisplay(mph) {
  assertFinite(mph, 'mph');
  return { value: mph, unit: 'mph' };
}

File: swift/Sources/FlightglassAdapter/Convert.swift
import FlightglassEngine

/// KONVERTERING — det ENE stedet enheter byttes (D57, D58, spec §11 krav 5).
///
/// Portert fra `adapter/src/convert.js`. Motoren regner i yards og mph og
/// røres aldri av enhetsvalget (D27).
///
/// D57: to enhetspakker — `meters` (avstander i meter) og `yards`. Fart er
/// ALLTID mph i begge. Enhetsuavhengig: spinn (rpm), smash (forholdstall),
/// vinkler (grader), Studios mm-verdier.
///
/// Konverteringsfaktoren er `Constants.yardToMetre` — aldri en lokal literal,
/// slik at faktoren ikke kan dupliseres (JS-siden håndhever det med lint).
public enum Convert {

  /// De to enhetspakkene fra D57. Interne id-er, ikke brukervendt kopi.
  public enum UnitSystem: String, CaseIterable, Sendable {
    case meters
    case yards
  }

  /// Kontraktsbrudd — spec §3: endelige tall inn, ellers kast.
  public enum ContractError: Error, Equatable {
    case nonFinite(name: String, value: Double)
  }

  public struct Distance: Equatable, Sendable {
    public let value: Double
    public let unit: String  // "m" | "yd"
  }

  public struct Speed: Equatable, Sendable {
    public let value: Double
    public let unit: String  // alltid "mph"
  }

  /// Avstand fra motor (yards) til visningsenhet. Returneres URUNDET —
  /// avrunding (D28) skjer i formatlaget, ETTER konvertering, aldri før.
  public static func distanceForDisplay(
    yards: Double, unitSystem: UnitSystem
  ) throws -> Distance {
    guard yards.isFinite else {
      throw ContractError.nonFinite(name: "yards", value: yards)
    }
    switch unitSystem {
    case .yards: return Distance(value: yards, unit: "yd")
    case .meters: return Distance(value: yards * Constants.yardToMetre, unit: "m")
    }
  }

  /// Fart. D57: alltid mph. Funksjonen finnes for at kallstedene skal gå
  /// gjennom konverteringslaget også der svaret er identitet — så regelen
  /// står ett sted, ikke implisitt i hvert kallsted.
  public static func speedForDisplay(mph: Double) throws -> Speed {
    guard mph.isFinite else {
      throw ContractError.nonFinite(name: "mph", value: mph)
    }
    return Speed(value: mph, unit: "mph")
  }
}
